src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexInBounds(index);
+
+        if (size == index) {
+            linkLast(value);
+        } else {
+            Node<T> nodeAtIndex = getNodeIndex(index);
+            Node<T> prevNode = nodeAtIndex.prev;
+            Node<T> newNode = new Node<>(prevNode, value, nodeAtIndex);
+            nodeAtIndex.prev = newNode;
+
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        return getNodeIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        Node<T> nodeAtIndex = getNodeIndex(index);
+        T changedNode = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return changedNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        checkIndexAtSizeRange(index);
+        return unlink(getNodeIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> nodeAtIndex = head; nodeAtIndex != null; nodeAtIndex = nodeAtIndex.next) {
+            if (nodeAtIndex.value == object || object != null && object.equals(nodeAtIndex.value)) {
+                unlink(nodeAtIndex);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexInBounds(int index) {
+        if (index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void checkIndexAtSizeRange(int index) {
+        if (index == size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
```

Comment:
One method is enough 
![image](https://user-images.githubusercontent.com/20326858/224541693-825db0e6-a76c-4d7e-a922-b513bb38292b.png)